### ISSUE TITLE
[Bug]: Nginx Module Can't Get data

### ISSUE THREAD

**Issue by alexhackney** [netdata/netdata#16432]
### Bug description

I'm trying to enable the nginx module to track connections and data in real time.

```
netdata@web-1:/usr/libexec/netdata/plugins.d$ ./go.d.plugin -d -m nginx
[ DEBUG ] main[main] main.go:120 plugin: name=go.d, version=v0.56.3
[ DEBUG ] main[main] main.go:122 current user: name=netdata, uid=116
[ INFO  ] main[main] main.go:126 env HTTP_PROXY '', HTTPS_PROXY ''
[ INFO  ] main[main] agent.go:145 instance is started
[ INFO  ] main[main] setup.go:21 loading config file
[ INFO  ] main[main] setup.go:29 looking for 'go.d.conf' in [/etc/netdata /usr/lib/netdata/conf.d]
[ INFO  ] main[main] setup.go:36 found '/usr/lib/netdata/conf.d/go.d.conf
[ INFO  ] main[main] setup.go:43 config successfully loaded
[ INFO  ] main[main] agent.go:149 using config: enabled 'true', default_run 'true', max_procs '0'
[ INFO  ] main[main] setup.go:48 loading modules
[ INFO  ] main[main] setup.go:67 enabled/registered modules: 1/80
[ INFO  ] main[main] setup.go:72 building discovery config
[ INFO  ] main[main] setup.go:119 looking for 'nginx.conf' in [/etc/netdata/go.d /usr/lib/netdata/conf.d/go.d]
[ INFO  ] main[main] setup.go:135 found '/etc/netdata/go.d/nginx.conf
[ INFO  ] main[main] setup.go:140 dummy/read/watch paths: 0/1/0
[ INFO  ] discovery[manager] manager.go:107 registered discoverers: [file discovery: [file reader]]
[ DEBUG ] functions[manager] manager.go:119 registering function 'get_plugin_config'
[ DEBUG ] functions[manager] manager.go:119 registering function 'get_plugin_config_schema'
[ DEBUG ] functions[manager] manager.go:119 registering function 'set_plugin_config'
[ DEBUG ] functions[manager] manager.go:119 registering function 'get_module_config'
[ DEBUG ] functions[manager] manager.go:119 registering function 'get_module_config_schema'
[ DEBUG ] functions[manager] manager.go:119 registering function 'set_module_config'
[ DEBUG ] functions[manager] manager.go:119 registering function 'get_job_config'
[ DEBUG ] functions[manager] manager.go:119 registering function 'get_job_config_schema'
[ DEBUG ] functions[manager] manager.go:119 registering function 'set_job_config'
[ DEBUG ] functions[manager] manager.go:119 registering function 'delete_job'
[ INFO  ] main[main] setup.go:154 looking for 'vnodes/' in [/etc/netdata /usr/lib/netdata/conf.d]
[ INFO  ] main[main] setup.go:166 found '/usr/lib/netdata/conf.d/vnodes' (0 vhosts)
[ INFO  ] discovery[manager] manager.go:61 instance is started
[ INFO  ] functions[manager] manager.go:47 instance is started
[ INFO  ] job[manager] manager.go:94 instance is started
[ INFO  ] discovery[file manager] discovery.go:63 instance is started
[ INFO  ] discovery[file reader] read.go:47 instance is started
[ INFO  ] discovery[file reader] read.go:48 instance is stopped
[ INFO  ] dyncfg[manager] dyncfg.go:65 instance is started
DYNCFG_ENABLE 'go.d'

[ DEBUG ] job[manager] manager.go:124 received config group ('/etc/netdata/go.d/nginx.conf'): 2 jobs (added: 2, removed: 0)
DYNCFG_REGISTER_MODULE 'nginx' job_array

DYNCFG_REGISTER_JOB 'nginx' 'local_13564587856247063018' 'stock' 0

[ DEBUG ] job[manager] manager.go:262 creating nginx[local] job, config: map[__provider__:file reader __source__:/etc/netdata/go.d/nginx.conf autodetection_retry:0 module:nginx name:local priority:70000 tls_skip_verify:true update_every:1 url:http://localhost/stub_status]
[ DEBUG ] nginx[local] nginx.go:76 using URL http://localhost/stub_status
[ DEBUG ] nginx[local] nginx.go:77 using timeout: 1s
[ ERROR ] nginx[local] nginx.go:93 error on request : Get "http://localhost/stub_status": EOF
[ ERROR ] nginx[local] job.go:204 check failed
REPORT_JOB_STATUS 'nginx' 'local_13564587856247063018' 'error' 0 'job detection failed, stopping it'

DYNCFG_REGISTER_JOB 'nginx' 'local_10774556093894337588' 'stock' 0

[ DEBUG ] job[manager] manager.go:262 creating nginx[local] job, config: map[__provider__:file reader __source__:/etc/netdata/go.d/nginx.conf autodetection_retry:0 module:nginx name:local priority:70000 update_every:1 url:http://127.0.0.1/stub_status]
[ DEBUG ] nginx[local] nginx.go:76 using URL http://127.0.0.1/stub_status
[ DEBUG ] nginx[local] nginx.go:77 using timeout: 1s
[ ERROR ] nginx[local] nginx.go:93 error on request : Get "http://127.0.0.1/stub_status": EOF
[ ERROR ] nginx[local] job.go:204 check failed
```

```
netdata@web:/usr/libexec/netdata/plugins.d$ curl http://localhost/stub_status
Active connections: 2
server accepts handled requests
 142 142 175
Reading: 0 Writing: 1 Waiting: 1
```

```
forge@web:/etc/nginx/conf.d$ cat status.conf
server {
  listen 80;
  server_name localhost;

  access_log on;
  allow 127.0.0.1;
  # deny all;

  location /stub_status {
    stub_status;
  }
}
```

### Expected behavior

I expect it to be able to get the data.

### Steps to reproduce

1. Install netdata on Ubuntu 22.04
2. Configure NGINX to provide the data.

### Installation method

kickstart.sh

### System info

```shell
Linux web 5.15.0-71-generic #78-Ubuntu SMP Tue Apr 18 09:00:29 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=22.04
/etc/lsb-release:DISTRIB_CODENAME=jammy
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"
/etc/os-release:PRETTY_NAME="Ubuntu 22.04.2 LTS"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION_ID="22.04"
/etc/os-release:VERSION="22.04.2 LTS (Jammy Jellyfish)"
/etc/os-release:VERSION_CODENAME=jammy
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:UBUNTU_CODENAME=jammy
```

### Netdata build info

```shell
Packaging:
    Netdata Version ____________________________________________ : v1.43.2
    Installation Type __________________________________________ : binpkg-deb
    Package Architecture _______________________________________ : x86_64
    Package Distro _____________________________________________ :
    Configure Options __________________________________________ :  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--disable-option-checking' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--prefix=/usr' '--sysconfdir=/etc' '--localstatedir=/var' '--libdir=/usr/lib' '--libexecdir=/usr/libexec' '--with-user=netdata' '--with-math' '--with-zlib' '--with-webdir=/var/lib/netdata/www' '--disable-dependency-tracking' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -ffile-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -ffile-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security'
Default Directories:
    User Configurations ________________________________________ : /etc/netdata
    Stock Configurations _______________________________________ : /usr/lib/netdata/conf.d
    Ephemeral Databases (metrics data, metadata) _______________ : /var/cache/netdata
    Permanent Databases ________________________________________ : /var/lib/netdata
    Plugins ____________________________________________________ : /usr/libexec/netdata/plugins.d
    Static Web Files ___________________________________________ : /var/lib/netdata/www
    Log Files __________________________________________________ : /var/log/netdata
    Lock Files _________________________________________________ : /var/lib/netdata/lock
    Home _______________________________________________________ : /var/lib/netdata
Operating System:
    Kernel _____________________________________________________ : Linux
    Kernel Version _____________________________________________ : 5.15.0-71-generic
    Operating System ___________________________________________ : Ubuntu
    Operating System ID ________________________________________ : ubuntu
    Operating System ID Like ___________________________________ : debian
    Operating System Version ___________________________________ : 22.04.2 LTS (Jammy Jellyfish)
    Operating System Version ID ________________________________ : none
    Detection __________________________________________________ : /etc/os-release
Hardware:
    CPU Cores __________________________________________________ : 1
    CPU Frequency ______________________________________________ : 3695000000
    CPU Architecture ___________________________________________ : 2071928832
    RAM Bytes __________________________________________________ : 68719476736
    Disk Capacity ______________________________________________ : x86_64
    Virtualization Technology __________________________________ : microsoft
    Virtualization Detection ___________________________________ : systemd-detect-virt
Container:
    Container __________________________________________________ : none
    Container Detection ________________________________________ : systemd-detect-virt
    Container Orchestrator _____________________________________ : none
    Container Operating System _________________________________ : none
    Container Operating System ID ______________________________ : none
    Container Operating System ID Like _________________________ : none
    Container Operating System Version _________________________ : none
    Container Operating System Version ID ______________________ : none
    Container Operating System Detection _______________________ : none
Features:
    Built For __________________________________________________ : Linux
    Netdata Cloud ______________________________________________ : YES
    Health (trigger alerts and send notifications) _____________ : YES
    Streaming (stream metrics to parent Netdata servers) _______ : YES
    Replication (fill the gaps of parent Netdata servers) ______ : YES
    Streaming and Replication Compression ______________________ : YES (lz4)
    Contexts (index all active and archived metrics) ___________ : YES
    Tiering (multiple dbs with different metrics resolution) ___ : YES (5)
    Machine Learning ___________________________________________ : YES
Database Engines:
    dbengine ___________________________________________________ : YES
    alloc ______________________________________________________ : YES
    ram ________________________________________________________ : YES
    map ________________________________________________________ : YES
    save _______________________________________________________ : YES
    none _______________________________________________________ : YES
Connectivity Capabilities:
    ACLK (Agent-Cloud Link: MQTT over WebSockets over TLS) _____ : YES
    static (Netdata internal web server) _______________________ : YES
    h2o (web server) ___________________________________________ : YES
    WebRTC (experimental) ______________________________________ : NO
    Native HTTPS (TLS Support) _________________________________ : YES
    TLS Host Verification ______________________________________ : YES
Libraries:
    LZ4 (extremely fast lossless compression algorithm) ________ : YES
    zlib (lossless data-compression library) ___________________ : YES
    Judy (high-performance dynamic arrays and hashtables) ______ : YES (bundled)
    dlib (robust machine learning toolkit) _____________________ : YES (bundled)
    protobuf (platform-neutral data serialization protocol) ____ : YES (system)
    OpenSSL (cryptography) _____________________________________ : YES
    libdatachannel (stand-alone WebRTC data channels) __________ : NO
    JSON-C (lightweight JSON manipulation) _____________________ : YES
    libcap (Linux capabilities system operations) ______________ : NO
    libcrypto (cryptographic functions) ________________________ : YES
    libm (mathematical functions) ______________________________ : YES
    jemalloc ___________________________________________________ : NO
    TCMalloc ___________________________________________________ : NO
Plugins:
    apps (monitor processes) ___________________________________ : YES
    cgroups (monitor containers and VMs) _______________________ : YES
    cgroup-network (associate interfaces to CGROUPS) ___________ : YES
    proc (monitor Linux systems) _______________________________ : YES
    tc (monitor Linux network QoS) _____________________________ : YES
    diskspace (monitor Linux mount points) _____________________ : YES
    freebsd (monitor FreeBSD systems) __________________________ : NO
    macos (monitor MacOS systems) ______________________________ : NO
    statsd (collect custom application metrics) ________________ : YES
    timex (check system clock synchronization) _________________ : YES
    idlejitter (check system latency and jitter) _______________ : YES
    bash (support shell data collection jobs - charts.d) _______ : YES
    debugfs (kernel debugging metrics) _________________________ : YES
    cups (monitor printers and print jobs) _____________________ : YES
    ebpf (monitor system calls) ________________________________ : YES
    freeipmi (monitor enterprise server H/W) ___________________ : YES
    nfacct (gather netfilter accounting) _______________________ : YES
    perf (collect kernel performance events) ___________________ : YES
    slabinfo (monitor kernel object caching) ___________________ : YES
    Xen ________________________________________________________ : NO
    Xen VBD Error Tracking _____________________________________ : NO
Exporters:
    AWS Kinesis ________________________________________________ : NO
    GCP PubSub _________________________________________________ : NO
    MongoDB ____________________________________________________ : YES
    Prometheus (OpenMetrics) Exporter __________________________ : YES
    Prometheus Remote Write ____________________________________ : YES
    Graphite ___________________________________________________ : YES
    Graphite HTTP / HTTPS ______________________________________ : YES
    JSON _______________________________________________________ : YES
    JSON HTTP / HTTPS __________________________________________ : YES
    OpenTSDB ___________________________________________________ : YES
    OpenTSDB HTTP / HTTPS ______________________________________ : YES
    All Metrics API ____________________________________________ : YES
    Shell (use metrics in shell scripts) _______________________ : YES
Debug/Developer Features:
    Trace All Netdata Allocations (with charts) ________________ : NO
    Developer Mode (more runtime checks, slower) _______________ : NO
```

**Comment by ilyam8** [comment]
> [ ERROR ] nginx[local] nginx.go:93 error on request : Get "http://localhost/stub_status": EOF

This error comes from golang std lib http [client.Do](https://pkg.go.dev/net/http#Client.Do) request. I can only assume that the connection is closed before the client can read headers or something.

I can't reproduce the problem using "Steps to reproduce".

1. show `curl -v localhost:80/stub_status`
2. check your web server access logs

**Comment by alexhackney** [comment]
As I mentioned I can't reproduce it on that machine even.

When I do the curl just in terminal: 
```
netdata@web:/usr/libexec/netdata/plugins.d$ curl http://localhost/stub_status
Active connections: 2
server accepts handled requests
 142 142 175
Reading: 0 Writing: 1 Waiting: 1
```

But netdata won't work.

**Comment by ilyam8** [comment]
I understood that. Unfortunately, I have no idea what the problem is. That is why I suggested to do 1. and 2. in my previous post.

**Comment by alexhackney** [comment]
I apologize, I misunderstood your message.

`curl -v localhost:80/stub_status
*   Trying 127.0.0.1:80...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /stub_status HTTP/1.1
> Host: localhost
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx/1.22.1
< Date: Mon, 20 Nov 2023 19:24:42 GMT
< Content-Type: text/plain
< Content-Length: 112
< Connection: keep-alive
<
Active connections: 51
server accepts handled requests
 63219 63219 539039
Reading: 0 Writing: 1 Waiting: 50
* Connection #0 to host localhost left intact`


There are no new messages in the access logs when I do this.

**Comment by ilyam8** [comment]
I believe this is not a Netdata or go.d.plugin problem, but a misconfiguration. I assume that in your `/etc/hosts` `localhost` is resolved as `127.0.0.1` and as `::1`. The difference between `curl` and `go.d.plugin` is:
 - `curl` prefers ipv4.
 - `go.d.plugin` prefers ipv6.


Adding `listen [::]:80;` to `status.conf` will fix the issue.
